### ISSUE TITLE
DR-DOS 3.40 / 3.41 failing to boot from fat16b filesystem with native boot block

### ISSUE THREAD

**Issue by andrewbird** [stsp/dosemu2#275]
I don't believe this a regression, but the following FAT16B boots fail on DR-DOS 3.40 & 3.41
~~~
FAIL: Test DR-DOS-3.40 FAT16B image file
FAIL: Test DR-DOS-3.40 FAT16B vfs directory boot.blk
FAIL: Test DR-DOS-3.41 FAT16B image file
FAIL: Test DR-DOS-3.41 FAT16B vfs directory boot.blk
~~~
Note this only occurs if the native boot block is used. Booting with dosemu boot block is fine

**Comment by stsp** [comment]
Isnt this fixed?

**Comment by andrewbird** [comment]
No, I'm looking at it now. If you are thinking about PR #273, that was FAT12 and FAT16 image file boot with DR-DOS 3.40 only

**Comment by andrewbird** [comment]
So, just to explain. At present DR-DOS 3.4x fails to boot with a FAT16B filesystem on our type 9 disk (900/15/17), either fatfs or mkfatimage16 when the native boot block has been used. Tweaking the following allows booting:
 - set cluster_secs initial value to 8 before loop
 - decrement fat_secs

A kludge patch below for fatfs.c (probably not correct for current HEAD) allows the boot to proceed, but obviously isn't the correct way to do it. I also did something similar for mkfatimage16.c which fixed that too. It's probably that the native boot block is a little naive and expects certain geometry etc. So I would leave it at that except that the geometry it's expecting matches that produced by mkfs.msdos, which leads me to conclude that perhaps dosemu is not doing quite the right thing.

```C
--- a/src/base/misc/fatfs.c
+++ b/src/base/misc/fatfs.c
@@ -200,15 +200,18 @@ void fatfs_init(struct disk *dp)
     f->fat_type = FAT_TYPE_FAT16;
     f->total_secs = num_sectors;
     f->root_secs = 32;
-    for (u = 4; u <= 512; u <<= 1) {
+    for (u = (num_sectors <= MAX_SECTORS_FAT16) ? 4 : 8; u <= 512; u <<= 1) {
       if (u * 0xfff0u > f->total_secs)
         break;
     }
     f->cluster_secs = u;
     f->fat_secs = ((f->total_secs / f->cluster_secs + 2) * 2 + 0x1ff) >> 9;
+    if (num_sectors > MAX_SECTORS_FAT16)
+      f->fat_secs--;
 
diff --git a/src/base/misc/fatfs.h b/src/base/misc/fatfs.h
index 737672e2..a5d2db82 100644
--- a/src/base/misc/fatfs.h
+++ b/src/base/misc/fatfs.h
@@ -10,6 +10,9 @@
 #define MAX_DIR_NAME_LEN       256     /* max size of fully qualified path */
 #define MAX_FILE_NAME_LEN      256     /* max size of a file name */
 
+#define MAX_SECTORS_FAT12 4078*8
+#define MAX_SECTORS_FAT16 0xffff
+
 typedef struct {
   struct {
     unsigned dir       :1;
```

**Comment by andrewbird** [comment]
I just copied this text here from the Disks08 PR for reference

Regarding the geometry here is a FAT16B BPB from DR-DOS 3.41
~~~
title: img
filename: test-imagedir/drives/c/boot.blk
oemname                         = "IBM  3.2"
bpb->bytes_per_sector           =     0x0200
bpb->sectors_per_cluster        =       0x08
bpb->reserved_sectors           =     0x0001
bpb->num_fats                   =       0x02
bpb->num_root_entries           =     0x0200
bpb->num_sectors_small          =     0x0000
bpb->media_type                 =       0xf8
bpb->sectors_per_fat            =     0x0070
bpb->sectors_per_track          =     0x0011
bpb->num_heads                  =     0x000f
~~~

and another dumped out during fatfs.c boot with same disk geometry

~~~
title: vfs
filename: /tmp/fatfs.blk
oemname                         = "DOSEMU10"
bpb->bytes_per_sector           =     0x0200
bpb->sectors_per_cluster        =       0x04
bpb->reserved_sectors           =     0x0001
bpb->num_fats                   =       0x02
bpb->num_root_entries           =     0x0200
bpb->num_sectors_small          =     0x0000
bpb->media_type                 =       0xf8
bpb->sectors_per_fat            =     0x00e1
bpb->sectors_per_track          =     0x0011
bpb->num_heads                  =     0x000f
~~~

As you can see, the cluster size is different (not necessarily wrong), but the resulting sectors_per_fat is off by one. With a sectors_per_cluster of 4 I think it should be 0xe0 not 0xe1. If I were to force the sectors_per_cluster to be 8 to match the DR-DOS one, the resulting sectors_per_fat would be 0x71, again one different.

**Comment by andrewbird** [comment]
I've taken the latest GPLv2+ version of the [mkdosfs.c](https://github.com/dosfstools/dosfstools/blob/61e7466965387a881671c74156a14659362c3a4e/src/mkdosfs.c) program and adapted its fat calculation algorithm. Attached is a test program that compares the existing with new method and writes them out. I'm pretty happy with the result as the majority of differences can be attributed to the fat_secs off by one I was looking for, but input would be appreciated before I write the patch for Dosemu?
[fat_calc.c](https://github.com/stsp/dosemu2/files/824931/fat_calc.c.txt)
[calc-result.txt](https://github.com/stsp/dosemu2/files/824930/calc-result.txt)

**Comment by stsp** [comment]
If it works fine, I wouldn't mind getting that in.
But please note that it is most likely a bug if someone
errors on just a harmless off-by-one increase of
bpb->sectors_per_fat. So you'll need to justify the
efforts by the good list of offenders. If this is just some
errant DOS version, _you_ should decide whether this
worth an efforts or not.

**Comment by andrewbird** [comment]
Yep, not worth integrating, as it breaks more boot tests than it fixes. I'm going to close and accept that really old versions of DR-DOS won't boot with fatfs/mkimage when using native boot block.

**Comment by stsp** [comment]
Wait, but why it breaks what?

**Comment by andrewbird** [comment]
On its own it didn't fix the DR-DOS 3.4x FAT16B issue, I had to increase the initial cluster size too from 4 to 8. That broke other DOSes. Since we have non-working MFS on DR-DOS < 7.0, it's of limited use even if a fix is found.
